packages/react/src/components/ComposedModal/ComposedModal.tsx

```diff
@@ -529,20 +533,33 @@ const ComposedModalDialog = React.forwardRef<
   useEffect(() => {
     if (!open) return;
 
-    const handleEscapeKey = (event) => {
+    const handleEscapeKey = (event: globalThis.KeyboardEvent) => {
       if (match(event, keys.Escape)) {
-        event.preventDefault();
-        event.stopPropagation();
-        closeModal(event);
+        const allModalContainers = document.querySelectorAll(
+          '[role="dialog"][aria-modal="true"]'
+        );
```

Comment:
For the original fix in https://github.com/carbon-design-system/carbon/pull/19908 did you look into the previous fix from https://github.com/carbon-design-system/carbon/pull/11609?

If we could rely on the `onKeyDown` on the `<Layer>` to fire appropriately, it would be scoped to only the current modal and so I think it would close only the topmost modal, one at a time. No need for the document event handler.